capstone-backend/src/main/java/com/google/sps/servlets/Utility.java

```diff
@@ -36,26 +44,185 @@ public static Firestore getFirestoreDb() throws IOException {
     return FirestoreClient.getFirestore();
   }
 
-  public static JsonObject createRequestBodyJson(HttpServletRequest request) {
-    JsonObject jsonObject = new JsonObject();
-    StringBuffer jb = new StringBuffer();
-    String line = null;
+  public static JsonObject createRequestBodyJson(HttpServletRequest request) throws IOException {
+    String jsonObjectString = request.getReader().lines().collect(Collectors.joining());
+    JsonObject jsonObject = JsonParser.parseString(jsonObjectString).getAsJsonObject();
+    return jsonObject;
+  }
+
+  /**
+   * Query the given collection by id. 
+   * @param {collectionReference} reference to the appropriate database collection
+   * @param {request} request sent to the backend
+   * @param {response} response returned from the call
+   * @param {genericClass} a Generic class, used in casting documents from the database
+   * @return List<T> a singleton List of the Object that matches the ID in the request.
+   */
+  private static <T> List<T> getById(CollectionReference collectionReference, HttpServletRequest request, 
+      HttpServletResponse response, GenericClass<T> genericClass) throws IOException {
+    if (request.getParameterMap().size() > 1) {
+      System.err.println("Error: No other parameter can be sent with an ID");
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return null;
+    }
+    List<T> retrievedObjects = new ArrayList<>();
+    DocumentReference docRef = collectionReference.document(request.getParameter("id"));
+    ApiFuture<DocumentSnapshot> asyncDocument = docRef.get();
+    DocumentSnapshot document = null;
+    T item = null;
     try {
-      BufferedReader reader = new BufferedReader(new InputStreamReader((request.getInputStream())));
-      while ((line = reader.readLine()) != null) {
-       jb.append(line);
+      document = asyncDocument.get();
+      if (document.exists()) {
+        item = document.toObject(genericClass.getMyType());
```

Comment:
Hey Steven -- I found out a solution to our id problem. We should be able to do document.getId() here and get the string id of the given data stored in firebase (https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/DocumentSnapshot). If we can ensure all of our objects in the java backend have setId methods, we could do item.setId() with the new id from the DocumentSnapshot object too!

This also applies to your other get sub-methods. It seems important to fit into this PR instead of waiting, since ids are important and we do need to access them on the frontend

Reply:
The problem is that since these methods are written for `Generic` types, there is no way to set the ID of the casted objects before returning them to the servlet. We would have to return a `List<QueryDocumentSnapshot>`, and then cast and set IDs in the servlet, before appending them to the response. This would introduce repeated code in the servlets, as they'd all need to perform this same functionality. 

After our discussion today, I think it would be best to use the Java UUID library instead, and have the primary index in firebase be the object's id set by the UUID functions. This will have some repeated data, as the object id and primary key will be the same, but the simplification of code outweighs this drawback in my mind.

Reply:
After further discussion, I think it's best to use Firebase's IDs. I created a class called BaseEntity that has the needed setId() function, and the generic functions apply to type <T extends BaseEntity> allowing us to set the ID of the generic objects retrieved before they are returned to the servlet. This way, the id is not stored as a field of the object in firebase, but only as the primary key. When it is returned to the frontend, it will have that primary key as its id.